Init the vaadin-button element with a native button

limonte opened the pull request:
Connected to https://github.com/vaadin/components-team-tasks/issues/231

Alternative of #1

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-button/2)
<!-- Reviewable:end -->

Saulis commented:
Review status: 0 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-button.html, line 36 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0vnJFcjqGKhGZ1B_G:-Kd0vnJGgo0yLzPz_8oL:be8g3yt) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L36)):*
> ```HTML
>     <button id="button"
>         role="button"
>         autofocus$="[[autofocus]]"
> ```

If we manually focus on `autoFocus` in `ready()` do we need to apply the property to the button?

---

*[vaadin-button.html, line 50 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uJ6ohuj8kQGQ6Mpv:-Kd0uJ6ohuj8kQGQ6Mpw:b-fgyfn2) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L50)):*
> ```HTML
>     if (!window.Vaadin) window.Vaadin = {};
> 
>     Vaadin.ButtonElement = class VaadinButtonElement extends Polymer.Element {
> ```

Why the global variable?

---

*[vaadin-button.html, line 63 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uttLycNGuse5mWCQ:-Kd0uttLycNGuse5mWCR:bltrxgn) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L63)):*
> ```HTML
>             autofocus: {
>               type: Boolean,
>               reflectToAttribute: true
> ```

Wouldn't expect this attribute to be used for styling, so no need to reflect it?

---

*[vaadin-button.html, line 104 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0wBywrKn-kqBDE0eB:-Kd0wBywrKn-kqBDE0eC:b-rdtyvp) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L104)):*
> ```HTML
> 
>         if (this.autofocus) {
>           this.$.button.focus();
> ```

This needs to be probably done inside `connectedCallback` – there's no guarantee that `this.$.button` is defined at this point yet. See https://www.polymer-project.org/2.0/docs/about_20#lifecycle-changes, test also in Firefox to see if it works on Polyfilled browsers.

---

*[vaadin-button.html, line 113 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0xTM8kdnvDRIn-heS:-Kd0xTM8kdnvDRIn-heT:b-7uko94) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L113)):*
> ```HTML
> 
>       _focus() {
>         if (this.disabled) {
> ```

I'm not sure if we should need to check this manually, `disabled` should/could make the button non-focusable to prevent the `focus` event from being fired in the first place.

---

*[vaadin-button.html, line 127 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0y9Jl7qPfnHQQmoUL:-Kd0y9Jl7qPfnHQQmoUM:bf54zm8) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L127)):*
> ```HTML
>         // shift-tab
>         if (e.shiftKey && e.keyCode === 9) {
>           this.focus();
> ```

in Shady DOM, this will just move the focus back to the native button, preventing the user from shift-tabbing out. You need to prevent focusing the native button when shift tabbing.

---

*[vaadin-button.html, line 134 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0x9vkGklUcjH08LVG:-Kd0x9vkGklUcjH08LVH:b20xpig) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L134)):*
> ```HTML
>         if (disabled) {
>           this._oldTabindex = this.tabindex;
>           this.tabindex = -1;
> ```

It might be more effective remove the whole `tabindex` attribute, which will effectively disable both also `focus()` and focus by clicking?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-button.html, line 36 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0vnJFcjqGKhGZ1B_G:-Kd1AztaoOHh5vIJ3Qc1:b-vuluz7) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L36)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

If we manually focus on `autoFocus` in `ready()` do we need to apply the property to the button?
</blockquote></details>

Probably for shady dom it would make sense. But given the case that FF does not respect it, proly we can safely remove this line

---

*[vaadin-button.html, line 50 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uJ6ohuj8kQGQ6Mpv:-Kd1BKpeMvzS03WZNBeo:b-iqqk2y) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L50)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Why the global variable?
</blockquote></details>

We reached some kind of agreement https://www.flowdock.com/app/vaadin/elements-team/threads/e54rD71FuILZFg4cfayOLVYuaGn 

---

*[vaadin-button.html, line 113 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0xTM8kdnvDRIn-heS:-Kd1BpjeAHMTdfmY3Jzm:brjzouw) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L113)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

I'm not sure if we should need to check this manually, `disabled` should/could make the button non-focusable to prevent the `focus` event from being fired in the first place.
</blockquote></details>

The container still continues being focusable at this point.

Anyway in the control-state mixin will correct all these issues. It will come in a separated PR, this is just a MVP to wrap a native button so I would continue merging it and then apply the mixin instead of trying to fix all these issues here.

Another option could be to wait until the mixin of the input has passed the review process, and apply to this PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-button.html, line 36 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0vnJFcjqGKhGZ1B_G:-Kd5IxQjqhq0AUC_ZyYl:b-p3sa38) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L36)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Probably for shady dom it would make sense. But given the case that FF does not respect it, proly we can safely remove this line
</blockquote></details>

Removed, thx!

---

*[vaadin-button.html, line 50 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uJ6ohuj8kQGQ6Mpv:-Kd5Izbnro76ubZEQHUo:bcad3jk) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L50)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

We reached some kind of agreement https://www.flowdock.com/app/vaadin/elements-team/threads/e54rD71FuILZFg4cfayOLVYuaGn 
</blockquote></details>

As Manolo said, this is what we agreed to do as I understand correctly. I followed the form-layout here: https://github.com/vaadin/vaadin-form-layout/pull/1/files#diff-ed406cf2bf9dbb9f9118bc7af3d3bb1d

---

*[vaadin-button.html, line 63 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uttLycNGuse5mWCQ:-Kd5JT1PLL6a0-XQy755:b-p3sa2v) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L63)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Wouldn't expect this attribute to be used for styling, so no need to reflect it?
</blockquote></details>

Removed, thx.

---

*[vaadin-button.html, line 104 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0wBywrKn-kqBDE0eB:-Kd5KHl-TvYwjXxPzUWz:b-896fix) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L104)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

This needs to be probably done inside `connectedCallback` – there's no guarantee that `this.$.button` is defined at this point yet. See https://www.polymer-project.org/2.0/docs/about_20#lifecycle-changes, test also in Firefox to see if it works on Polyfilled browsers.
</blockquote></details>

Done.

---

*[vaadin-button.html, line 113 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0xTM8kdnvDRIn-heS:-Kd5vwZAoQjexEAa_Xtt:bss4a2b) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L113)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

The container still continues being focusable at this point.

Anyway in the control-state mixin will correct all these issues. It will come in a separated PR, this is just a MVP to wrap a native button so I would continue merging it and then apply the mixin instead of trying to fix all these issues here.

Another option could be to wait until the mixin of the input has passed the review process, and apply to this PR
</blockquote></details>

Removed.

---

*[vaadin-button.html, line 127 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0y9Jl7qPfnHQQmoUL:-Kd5vy64igXkjdjbZKw3:b-5rbivn) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L127)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

in Shady DOM, this will just move the focus back to the native button, preventing the user from shift-tabbing out. You need to prevent focusing the native button when shift tabbing.
</blockquote></details>

Tabbing will be handled in the next PR where I will replace the common stuff with the mixin Manolo is working right now

---

*[vaadin-button.html, line 134 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0x9vkGklUcjH08LVG:-Kd5w7z3uIgbmXjVC__o:b-5rbivn) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L134)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

It might be more effective remove the whole `tabindex` attribute, which will effectively disable both also `focus()` and focus by clicking?
</blockquote></details>

Tabbing will be handled in the next PR where I will replace the common stuff with the mixin Manolo is working right now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
It's not necessary to duplicate the code since we can depend on input, or vaadin-common when we have mixin moved there in case.

This PR depends on another one in vaadin-input that makes the test available for using in any component, and includes the autofocus property.

---

Review status: 0 of 26 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-button.html, line 50 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0uJ6ohuj8kQGQ6Mpv:-KdBzBkaYljP6EeMVz8u:b98df5t) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L50)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

As Manolo said, this is what we agreed to do as I understand correctly. I followed the form-layout here: https://github.com/vaadin/vaadin-form-layout/pull/1/files#diff-ed406cf2bf9dbb9f9118bc7af3d3bb1d
</blockquote></details>

We don't need this global variable. And also Hydrolysis is unable to produce any API doc if we have it, so removed from here as it was removed in vaadin-input.

---

*[vaadin-button.html, line 127 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0y9Jl7qPfnHQQmoUL:-KdBzb86OILwbCsKOPUW:bg25z2n) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L127)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Tabbing will be handled in the next PR where I will replace the common stuff with the mixin Manolo is working right now
</blockquote></details>

Handled in the  extended control-status mixin. Done.

---

*[vaadin-button.html, line 134 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kd0x9vkGklUcjH08LVG:-KdBzSNPH7Za3krFcEkj:bwlh4cf) ([raw file](https://github.com/vaadin/vaadin-button/blob/3571566f4aa3dacd8fbae6581f8b15f55ac651a1/vaadin-button.html#L134)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Tabbing will be handled in the next PR where I will replace the common stuff with the mixin Manolo is working right now
</blockquote></details>

Handled in the  extended control-status mixin. Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-:-KdByiMnq7OfVUdDhARP:buo0drq)*
<!-- Sent from Reviewable.io -->

jouni commented on vaadin-button.html:
```diff
@@ -17,30 +17,39 @@
       }
     </style>
+    <button id="button"
+        role="button"
```
This is most likely redundant. The native button element should already be announced as a button. You don’t need this when using the button element on a normal website, so why would you need it inside shadow DOM?

manolo commented:
Review status: 0 of 20 files reviewed at latest revision, 8 unresolved discussions.

---

*[vaadin-button.html, line 36 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-Kcx2CVsmYXWiUl2FuDp-r3-36:-KdDDDlUDFTQgPu-vc-q:bb6un1j) ([raw file](https://github.com/vaadin/vaadin-button/blob/2f73af6651b4aa23de4306a494bc3beb0b1e89e3/vaadin-button.html#L36)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary><blockquote>

This is most likely redundant. The native button element should already be announced as a button. You don’t need this when using the button element on a normal website, so why would you need it inside shadow DOM?
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 8 of 32 files at r1, 1 of 12 files at r2, 1 of 5 files at r3, 11 of 21 files at r4, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-button/2#-:-KdW4ElixLgm3UyMVM_j:b-uwkowd)*
<!-- Sent from Reviewable.io -->
